zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs

```diff
@@ -126,31 +124,40 @@ impl InputSource for TransactionRecordsById {
         anchor_height: zcash_primitives::consensus::BlockHeight,
         exclude: &[Self::NoteRef],
     ) -> Result<SpendableNotes<Self::NoteRef>, Self::Error> {
-        let mut unselected =
-            self.get_spendable_note_ids_and_values(sources, anchor_height, exclude);
-
+        let mut unselected = self
+            .get_spendable_note_ids_and_values(sources, anchor_height, exclude)
+            .into_iter()
+            .map(|(id, value)| NonNegativeAmount::from_u64(value).map(|value| (id, value)))
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(InputSourceError::InvalidValue)?;
         unselected.sort_by_key(|(_id, value)| *value);
-
         let mut selected = vec![];
-
         let mut index_of_unselected = 0;
+
         loop {
+            // if no unselected notes are available, return the currently selected notes even if the target value has not been reached
             if unselected.is_empty() {
-                // all notes are selected. we pass the max value onwards whether we have reached target or not
+                break;
+            }
+
+            // update target value for further note selection
+            let selected_notes_total_value = selected
+                .iter()
+                .try_fold(NonNegativeAmount::ZERO, |acc, (_id, value)| acc + *value)
+                .ok_or(InputSourceError::InvalidValue(BalanceError::Overflow))?;
+            let Some(updated_target_value) = target_value - selected_notes_total_value else {
+                // if underflow, target has been reached
```

Comment:
The underflow case is represented by `None` here:

https://github.com/zingolabs/librustzcash/blob/4c67b473cf1acc45edf5619e2862f2f3f8775a7a/components/zcash_protocol/src/value.rs#L353


not to be confused with a BalanceError::Underflow...


But the doc-comment for the BalanceError indicates that it's defined for exactly this case:

https://github.com/zingolabs/librustzcash/blob/4c67b473cf1acc45edf5619e2862f2f3f8775a7a/components/zcash_protocol/src/value.rs#L394

This seems inconsistent to me.

Reply:
yeah thats correct, as it fails in only one case I don't see it worthy or a PR into LRZ but i agree